### DISCUSSION TITLE
useForm extensibility

### DISCUSSION THREAD

**Discussion by chilltemp** [mantinedev/mantine#614] [Feedback]
Thoughts on making `useForm()` easier to extend?  My use case is Zod.js.  I use it for schema design and object validation.  Since it already knows how to validate my objects, I want to integrate it with useForm.  I'm sure that others have there preferred schema/validation library.  If you can ..., then it would make it easier for myself and others to wrap useForm with our own hook:
* export all of the interfaces from use-form.ts (some are exported from the file, but not the hooks library)
* add `overrideValidation` as an optional property of `UseFormInput`
* modify `validate` so that it invokes `overrideValidation`

How I'm accomplishing it today:
```typescript
import { useForm } from '@mantine/hooks';
import React, { ReactNode, useState } from 'react';
import { ZodType } from 'zod';

/* BEGIN: Copied from https://github.com/mantinedev/mantine/blob/master/src/mantine-hooks/src/use-form/use-form.ts
type ValidationRule<T> = {
type UseFormErrors<T> = {
interface UseFormInput<T> {
interface UseForm<T> {
// END copied **********************************************/

type ErrorFormatter = (errors: string[]) => ReactNode;

export interface UseZodFormInput<T> extends UseFormInput<T> {
  schema: ZodType<T>;
  errorFormatter?: ErrorFormatter;
}

export interface UseZodForm<T> extends UseForm<T> {
  formErrors?: ReactNode;
  setFormErrors: (errors: ReactNode | undefined) => void;
}

export function useZodForm<T extends { [key: string]: any }>({
  schema,
  errorFormatter,
  validationRules,
  errorMessages,
  initialValues,
}: UseZodFormInput<T>): UseZodForm<T> {
  const {
    values,
    errors,
    // validate,
    reset,
    setErrors,
    setValues,
    setFieldValue,
    setFieldError,
    validateField,
    resetErrors,
    // onSubmit,
    getInputProps,
  } = useForm({ validationRules, errorMessages, initialValues });

  const [formErrors, setFormErrors] = useState<ReactNode | undefined>(
    undefined
  );
  // Copied from useForm
  type ValidationErrors = Record<keyof T, React.ReactNode>;

  // Convert Zod errors into ReactNode, with simple default
  const formatter = errorFormatter || ((errors) => errors.join('; '));

  // This could be part of UseFormInput.  Allowing the wrapper hook to do additional validation.
  // This could append to existing errors, or completely replace them.
  // The code shown, appends the errors.  Allowing the default behavior of validationRules & errorMessages.
  const overrideValidation = (
    isValid: boolean,
    validationErrors: ValidationErrors
  ) => {
    const parsed = schema.safeParse(values);

    if (!parsed.success) {
      const errors = parsed.error.flatten();

      setFormErrors(
        !!errors.formErrors.length ? formatter(errors.formErrors) : undefined
      );

      // Merge existing and Zod field errors
      for (const [name, value] of Object.entries(errors.fieldErrors)) {
        const key = name as keyof T;

        validationErrors[key] = !!validationErrors[key] ? (
          // Use a Fragment to merge the errors from differing sources.
          // This assumes that errorMessages & errorFormatter will have similar styles.
          <>
            {validationErrors[key]}
            {formatter(value)}
          </>
        ) : (
          formatter(value)
        );
      }
    }

    return {
      isValid: isValid && parsed.success,
      validationErrors,
    };
  };

  // Original validate function with minor additions
  const validate = () => {
    let isValid = true;

    // Changed: from const to let
    let validationErrors = Object.keys(values).reduce((acc, field) => {
      if (
        validationRules &&
        typeof validationRules[field] === 'function' &&
        // Added: ?. because object could be undefined
        !validationRules[field]?.(values[field], values)
      ) {
        // Added: ?. because object could be undefined
        acc[field as keyof T] = errorMessages?.[field] || true;
        isValid = false;
      } else {
        acc[field as keyof T] = null;
      }

      return acc;
    }, {} as ValidationErrors);

    // Added: Call additional validation.  Pretending it's optional, as it would be if this code was in the original hook
    if (!!overrideValidation) {
      const additional = overrideValidation(isValid, validationErrors);
      isValid = additional.isValid;
      validationErrors = additional.validationErrors;
    }

    setErrors(validationErrors);
    return isValid;
  };

  // Have to completely replace the original onSubmit so that it can call the new validate fn.
  // No changes would be needed to onSubmit if overrideValidation() was implemented in validate().
  const onSubmit =
    (handleSubmit: (values: T) => any) => (event?: React.FormEvent) => {
      event && event.preventDefault();
      validate() && handleSubmit(values);
    };

  return {
    values,
    errors,
    validate,
    reset,
    setErrors,
    setValues,
    setFieldValue,
    setFieldError,
    validateField,
    resetErrors,
    onSubmit,
    getInputProps,
    // Zod can report errors that aren't attached to a specific field
    formErrors,
    setFormErrors,
  };
}
```

**Comment by MartinMuzatko** [comment]
I would very much favor such a mechanism, although I can't see how you leverage `errorMessages` to make use of zods dynamic errors.

Integrating Mantine with other form management libraries like formik or final-form still seems less favorable than using mantines built-ins. It would at least involve even more boilerplate and adapters to work properly. Mantine is doing a great job here with `getInputProps` which is well designed. But for me, Mantines form validation mechanisms are still lacking that major part you are describing. 

I would quite much prefer an approach as outlined in this issue: https://github.com/mantinedev/mantine/issues/533
The validation is fine as it is, but the error handling could be more verbose and allow for dynamic error texts. In that matter, I think Mantine would benefit from combining both the `validationRules` and `errorMessages`. There is no need for `validationRules` to return a `boolean` to indicate that the error message needs to display, when you can use a `string` as error message and no return value (`undefined`) or false at all to indicate no error.

**Reply by chilltemp** [reply]
I'm not really using `validationRules` or `errorMessages` in the above code.  I wrote `overrideValidation()` in a way that it will merge useForm's built in handling of `validationRules` and `errorMessages`, with the results from zod's `safeParse()`.  The above POC was intended to be a non-breaking change.

Agreed.  I love `getInputProps`, and am enhancing it in my wrapper hook.  My scenario for that enhancement is to handle date serialization.  I store dates in the db as strings in YYYY-MM-DD format.  So I've added `getInputProps(___, { type: 'date' })` to handle this scenario, but thats probably too custom to be included in useForm.

I agree with #553, it will allow for more dynamic errors right "out of the box".  Since zod does all the validation at once, it would still need another way to integrate (unless I missed something in that ticket).  The other thing about zod is that it can return errors that aren't field specific, thus the addition of `formErrors` and `setFormErrors` in my POC.

It sounds like @rtivital may be considering breaking changes to support extensibility (and #553).  If so, that open a lot of options I haven't considered.

**Reply by chilltemp** [reply]
Here's the simplified validation function:
```typescript
  const validate = () => {
    const parsed = schema.safeParse(values);
    const newErrors = !parsed.success ? parsed.error.flatten() : undefined;

    const newFormErrors: typeof formErrors = !!newErrors?.formErrors.length
      ? formatter(newErrors.formErrors)
      : null;

    const newFieldErrors: typeof errors = Object.keys(values).reduce(
      (acc, field) => {
        acc[field as keyof T] = !!newErrors?.fieldErrors[field]
          ? formatter(newErrors?.fieldErrors[field])
          : null;

        return acc;
      },
      {} as ValidationErrors
    );

    setErrors(newFieldErrors);
    setFormErrors(newFormErrors);
    return parsed.success;
  };
```

**Comment by rtivital** [comment]
Thanks for feedback, I'll create new package @mantine/form in one of the future releases which will support new features like schema validation with zod.

**Reply by chilltemp** [reply]
Awesome!  Thanks!

**Reply by kasfulk** [reply]
Awesome! We are still waiting for these features.

**Reply by rtivital** [reply]
`@mantine/form` package is available for a while now, it has schema validation support.

**Reply by ereburg** [reply]
Is it any chance to see such implementation with superstruct.js? I didn't find a guide how to add validation scheme manually. 
Thanks in advance!

**Reply by rtivital** [reply]
You can build your own resolver for any schema library, use other resolvers as reference
